### ISSUE TITLE
C-bill costs not displaying properly

### ISSUE THREAD

**Issue by Maelwys** [Solaris-Skunk-Werks/solarisskunkwerks#55]
It took me a bit to figure it out, but SSW seems to be having trouble with the Final C-bill cost calculation, or maybe just the display.  For instance. the HSN-10G Hellspawn

Item                                            DefBV     OffBV             Cost
Internal Structure - Endo Steel                 56         0           72,000.00
Engine - XL Fusion Engine                        0         0        3,240,000.00
Gyro - Gyro                                     23         0          900,000.00
Cockpit - Standard                               0         0          340,000.00
Heat Sinks - Double Heat Sink                    0         0           60,000.00
Musculature - Standard                           0        45           90,000.00
Actuators                                        0         0           45,000.00
Jump Jets - Standard Jump Jet                    0         0          225,000.00
Armor - Light Ferro-Fibrous                    318         0          112,500.00

Medium Pulse Laser                                  0        48        60,000.00
Medium Pulse Laser                                  0        48        60,000.00
MML-7                                               0        67       105,000.00
MML-7                                               0        67       105,000.00
Light PPC                                           0        88       150,000.00
@ MML-7 (SRM)                                       0         8        27,000.00
@ MML-7 (LRM)                                       0         8        30,000.00
@ MML-7 (LRM)                                       0         8        30,000.00

Cost Multiplier                                                            1.450
Dry Cost                                                               8,068,525
Total Cost                                                             8,068,525

As you can see, the Dry Cost and the Total Cost are the same. However, the Total Cost should be the Dry Cost plus the basic value of the ammunition.

A little note. Page 282 of TM states "As a general guide, when using these figures to compute a unit's final cost, it is not necessary to purchase ammunition for ammo-based weaponry." SSW follows this and doesn't include the cost of ammo when figuring out the Dry Cost of the design, but then is supposed to add it in to get the total cost.

This is a big difference between SSW and MegaMek. MM DOES include the ammo as part of the basic cost, and thus its multiplied by the Mech's cost modifier, so straight comparisons between the two programs isn't easy.

**Comment by WEKarnesky** [comment]
I would like to confirm that the official record sheets include ammo costs before changing this.

If they don't include costs, then I think we should examine whether we want to follow in MM's footsteps or not before actually implementing this change.

**Comment by Notsonoble** [comment]
Official Records Sheets dropped cost altogether a while back. None of the 2011 RS books have prices.

TROs I think are dry cost.

**Comment by Maelwys** [comment]
The newest sheets don't show the cost. And SSW shouldn't follow MM's footsteps in the case of the cost because MM is doing it wrong.

SSW is doing it mostly right. As far as I can tell, the Dry Cost is correct.

However, it is not calculating (or displaying) the Total Cost Correctly.

The Total Cost should be

Dry Cost + Basic Ammo prices.

Right now, SSW is just showing the Dry cost for the Total Cost

**Comment by Notsonoble** [comment]
I also just confirmed that the MUL is using dry cost. Although, they round up to a full c-bill while SSW actually rounds based on typical mathematical (significant figures) rules.

**Comment by WEKarnesky** [comment]
So we need to round up to the nearest integer.

For the output, what is the preference:

1. Cost: X,000,000 C-Bills (+Y,YYY in ammo)
* Total Cost: Z,000,000 C-Bills
* Dry Cost (No Ammo): X,000,000 C-Bills
* Something else?

**Comment by Notsonoble** [comment]
If cost is on record sheets, I like option 1. Especially since it shows accurately for your specialty ammo. 
For TRO output I think:
....
Dry Cost: X,000,000
Total Cost: Z,000,000 in this configuration.
....
would be the most accurate.

**Comment by Maelwys** [comment]
Yeah, I think Notsonoble is right on that, and for the TRO its already like that, just not showing correctly, so no need to change the formatting or anything.

**Comment by WEKarnesky** [comment]
Okay, 

Recordsheet will be updated to `Cost: X,000,000 C-Bills (+Y,YYY in ammo)`
TRO output will be updated to 
> Dry Cost: X,000,000
> Total Cost: Z,000,000 in this configuration.

Sounds good

**Comment by WEKarnesky** [comment]
Components/Mech.Java, Line 3166 is the mechs cost. 

https://github.com/Solaris-Skunk-Werks/solarisskunkwerks/blob/cb5ceb6831f91e95a234b2b50304b4bcf67bcbd7/sswlib/src/main/java/components/Mech.java#L3166-L3182

 Seems the ticket is incorrect and Also including ammo cost in both.  And I'm not sure if the cost is done correctly in the first place, here is the referenced function.

https://github.com/Solaris-Skunk-Werks/solarisskunkwerks/blob/cb5ceb6831f91e95a234b2b50304b4bcf67bcbd7/sswlib/src/main/java/components/Mech.java#L3243-L3274

  It is also including ammo cost in the multiplier and I'm not sure that is correct, going to look into the specific cost rules so I can cite the page number before we fix.

TL;DR:  I'm not sure ***anyone*** is doing costs correctly and we probably need to double-check at least this part.

**Comment by WEKarnesky** [comment]
okay, important note for Rounding from Page 274 of TechManual:

```
Rounding
All costs computed in TechManual must not be rounded until
the end of the calculations process, after applying any fnal cost
multipliers. At that point, costs should be rounded up to the
nearest .01 C-bill.
In the earlier stages of calculations, however, designers may
either leave all values unrounded, or—for more manageable
numbers—rounded normally to three signifcant digits after the
decimal point (thus, a value of 1.4563 could be rounded to 1.456
with minimal impact on computations)
```

So rounding Up to the nearest C-Bill will not be done.

**Comment by WEKarnesky** [comment]
After discord conversations, I believe that we have agreed to change 
```java
 public double GetTotalCost() { 
     // final cost calculations 
     return ( GetEquipCost() + GetChassisCost() ) * GetCostMult(); 
 } 
```

To

```java
 public double GetTotalCost() { 
     // final cost calculations 
     return (( GetEquipCost() + GetChassisCost() ) * GetCostMult()) + GetAmmoCosts() ; 
 } 
```

At that point, I'm not sure any printing output needs to be changed as they reference the two functions properly unless we want the ammo breakdown.

**Comment by Maelwys** [comment]
Alright. Should save me from having to pull out excel, since it was working as shown earlier but that might have been one of the non-public changes.

**Comment by WEKarnesky** [comment]
I'm not the one coding, but this is what they will look like when the PR is ready:

Record Sheet:

https://cdn.discordapp.com/attachments/601060598771482632/629496823379394570/unknown.png

TRO Output:

```
Dry Cost: 8,068,525 C-Bills
Total Cost: 8,194,675 C-Bills
```
